### PR TITLE
Added ppc64le support

### PR DESCRIPTION
**Pull request by mayurwaghmode**
This PR will add ppc64le support to the Cachi2 image.

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- n/a Code coverage from testing does not decrease and new code is covered
- n/a New code has type annotations
- n/a Docs updated (if applicable)
- n/a Docs links in the code are still valid (if docs were updated)
- n/a [Issue](https://github.com/containerbuildsystem/cachi2/issues/13) that tracks breaking changes to the OSBS2 integration is updated (if applicable)

### CODE REVIEW

**Comment by chmeliik** [comment]
Who needs this and what for? Could you explain that in the commit message?

**Review by chmeliik — COMMENTED** [review]
Seems reasonable, thanks for the contribution :+1: 

Would it be worth mentioning in the docs which platforms our container image supports? https://github.com/containerbuildsystem/cachi2#container-image. Although the quay.io UI should show that, maybe we don't need it

**Comment by chmeliik** [comment]
I was looking at the CI and noticed that the build now takes >10m, as opposed to <2m before

https://github.com/containerbuildsystem/cachi2/actions/runs/3675322031/jobs/6215005901

https://github.com/containerbuildsystem/cachi2/actions/runs/3649655142/jobs/6164612441

But I don't suppose there's anything we could do about that :disappointed:. The only idea I had was to not build the ppc64le image on every PR, but only when merging to main. But that might mean we miss issues in the build and it will fail once merged to main :-/

**Comment by mayurwaghmode** [comment]
> Seems reasonable, thanks for the contribution 👍
> 
> Would it be worth mentioning in the docs which platforms our container image supports? https://github.com/containerbuildsystem/cachi2#container-image. Although the quay.io UI should show that, maybe we don't need it

yeah, I also feel the same. Supported architecture details can be fetched from quay.io.

**Review by chmeliik — APPROVED** [review]
LGTM :+1:

**Comment by brunoapimentel** [comment]
> I was looking at the CI and noticed that the build now takes >10m, as opposed to <2m before
> 
> https://github.com/containerbuildsystem/cachi2/actions/runs/3675322031/jobs/6215005901
> 
> https://github.com/containerbuildsystem/cachi2/actions/runs/3649655142/jobs/6164612441
> 
> But I don't suppose there's anything we could do about that disappointed. The only idea I had was to not build the ppc64le image on every PR, but only when merging to main. But that might mean we miss issues in the build and it will fail once merged to main :-/

Can we start by only building the ppc65le image when merging to main? We can adjust this later if we feel like we're actually missing issues. I feel like the 5x longer CI will impact us more right now.

**Comment by chmeliik** [comment]
> Can we start by only building the ppc65le image when merging to main? We can adjust this later if we feel like we're actually missing issues. I feel like the 5x longer CI will impact us more right now.

Good point. And it's not like we actually run the "Smoketest" step on the ppc64 image anyway, we implicitly test only the x86 one (I think?)

I'd +1 that we can build ppc64 only on merges to main

**Comment by mayurwaghmode** [comment]
> > Can we start by only building the ppc65le image when merging to main? We can adjust this later if we feel like we're actually missing issues. I feel like the 5x longer CI will impact us more right now.
> 
> Good point. And it's not like we actually run the "Smoketest" step on the ppc64 image anyway, we implicitly test only the x86 one (I think?)
> 
> I'd +1 that we can build ppc64 only on merges to main

@brunoapimentel can we go ahead with this approach?

**Comment by brunoapimentel** [comment]
> > I'd +1 that we can build ppc64 only on merges to main
> 
> @brunoapimentel can we go ahead with this approach?
Sure, I agree with that!

Can you update the CI to only perform ppc65le builds on merges to main, then? We can merge after that.

**Review by lkolacek — APPROVED** [review]
Looks good! +1 for pushing only on merge to main. Thanks for contribution. :+1:

**Comment by chmeliik** [comment]
How do we want to do the conditional ppc64 builds? I can think of 2 options:
1. split the [build-image](https://github.com/containerbuildsystem/cachi2/blob/267895291b84fecebbc188ca15223216b0992b5e/.github/workflows/gating.yaml#L104) step into two
    * integration-test: build an x86 image, run integration tests on it
    * build-image: build a multiarch image, push it quay.io
      * only if integration-test succeeds, only on merges to `main`
2. same as above but put the multiarch build in a completely separate workflow
    * we can run the workflow separately if needed
    * the "only on merges to main" condition will be cleaner
    * but it will no longer depend on the success of integration tests
      *  although anything that gets merged to main was already integration-tested on the PR

**Comment by chmeliik** [comment]
> How do we want to do the conditional ppc64 builds? I can think of 2 options:
> 
>     1. split the [build-image](https://github.com/containerbuildsystem/cachi2/blob/267895291b84fecebbc188ca15223216b0992b5e/.github/workflows/gating.yaml#L104) step into two
>        
>        * integration-test: build an x86 image, run integration tests on it, don't push it anywhere
>        * build-image: build a multiarch image, push it quay.io
>          
>          * only if integration-test succeeds, only on merges to `main`
> 
>     2. same as above but put the multiarch build in a completely separate workflow
>        
>        * we can run the workflow separately if needed
>        * the "only on merges to main" condition will be cleaner
>        * but it will no longer depend on the success of integration tests
>          
>          * although anything that gets merged to main was already integration-tested on the PR

I would vote for option 2. @brunoapimentel @lkolacek thoughts?

**Comment by brunoapimentel** [comment]
> > How do we want to do the conditional ppc64 builds? I can think of 2 options:
> > ```
> > 1. split the [build-image](https://github.com/containerbuildsystem/cachi2/blob/267895291b84fecebbc188ca15223216b0992b5e/.github/workflows/gating.yaml#L104) step into two
> >    
> >    * integration-test: build an x86 image, run integration tests on it, don't push it anywhere
> >    * build-image: build a multiarch image, push it quay.io
> >      
> >      * only if integration-test succeeds, only on merges to `main`
> > 
> > 2. same as above but put the multiarch build in a completely separate workflow
> >    
> >    * we can run the workflow separately if needed
> >    * the "only on merges to main" condition will be cleaner
> >    * but it will no longer depend on the success of integration tests
> >      
> >      * although anything that gets merged to main was already integration-tested on the PR
> > ```
> 
> I would vote for option 2. @brunoapimentel @lkolacek thoughts?

Yeah, I agree. 2 seems cleaner.

**Comment by lkolacek** [comment]
> How do we want to do the conditional ppc64 builds? I can think of 2 options:

Sorry guys for the late response here. Yeah, approach 2 seems better. That build and test CI pipeline is starting to be too long anyway. Leaving the testing only to PR should be safe.

**Comment by mayurwaghmode** [comment]
> How do we want to do the conditional ppc64 builds? I can think of 2 options:
> 
> 1. split the [build-image](https://github.com/containerbuildsystem/cachi2/blob/267895291b84fecebbc188ca15223216b0992b5e/.github/workflows/gating.yaml#L104) step into two
>    
>    * integration-test: build an x86 image, run integration tests on it, don't push it anywhere
>    * build-image: build a multiarch image, push it quay.io
>      
>      * only if integration-test succeeds, only on merges to `main`
> 2. same as above but put the multiarch build in a completely separate workflow
>    
>    * we can run the workflow separately if needed
>    * the "only on merges to main" condition will be cleaner
>    * but it will no longer depend on the success of integration tests
>      
>      * although anything that gets merged to main was already integration-tested on the PR

@chmeliik I have created a separate workflow for building and publishing the multi-arch image. Also from the existing workflow have removed the single architecture image push part.  Can u please review it?

File: .github/workflows/multi-arch-gating.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            cachi2:${{ github.sha }} \
+            fetch-deps \
+            --source /tmp/source \
+            --output ./cachi2-output \
+            --package '{"path": ".", "type": "gomod"}'
+      - name: Run integration tests on built image
+        env:
+          CACHI2_IMAGE: localhost/cachi2:${{ github.sha }}
+        run: |
+           tox -e integration
```
Comment (chmeliik) [inline .github/workflows/multi-arch-gating.yaml:50]: I think the idea was for this workflow to be _just_ an image build, no tests necessary. I would drop integration tests from here for sure, even the smoketest is probably not necessary

File: .github/workflows/multi-arch-gating.yaml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      - name: Podman Login
+        uses: redhat-actions/podman-login@v1
+        if: github.ref == 'refs/heads/main'
+        with:
+          registry: quay.io
+          username: ${{ secrets.QUAY_USER }}
+          password: ${{ secrets.QUAY_PWD }}
+
+      - name: Push image to Quay.io registry
+        if: github.ref == 'refs/heads/main'
```
Comment (chmeliik) [inline .github/workflows/multi-arch-gating.yaml:61]: the `if` conditions are not needed anymore

File: .github/workflows/multi-arch-gating.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
+name: Mult-Arch-Gating
+
+on:
+  push:
+    branches:
+      - main
+
+jobs:
+  build-image:
+    name: Build Cachi2 image and run smoketest on it
```
Comment (chmeliik) [inline .github/workflows/multi-arch-gating.yaml:10]: If we drop the tests the names should be updated accordingly

File: .github/workflows/multi-arch-gating.yaml (hunk 1 above)
Comment (mayurwaghmode) [reply .github/workflows/multi-arch-gating.yaml]: I have removed the tests. @chmeliik can u check it?

File: .github/workflows/multi-arch-gating.yaml (hunk 3 above)
Comment (mayurwaghmode) [reply .github/workflows/multi-arch-gating.yaml]: Have renamed the workflow with multi-arch-build

**Review by chmeliik — APPROVED** [review]
LGTM :+1:

**Review by brunoapimentel — APPROVED** [review]

**Pull request merged**